File: backend/src/match_engine.cpp
#include "models.hpp"

#include <algorithm>
#include <cmath>
#include <map>
#include <random>
#include <sstream>

namespace fm {
namespace {

static void add_event(MatchOutput& output, int minute, const std::string& event_type, const std::string& team_name,
                      const std::string& player_name, const std::string& assist_name, const std::string& commentary) {
    output.events.push_back(MatchEvent{minute, event_type, team_name, player_name, assist_name, commentary});
}

static double player_overall(const PlayerSummary& p) {
    if (p.position == "Goalkeeper") {
        return p.goalkeeping * 0.45 + p.physical * 0.2 + p.passing * 0.1 + p.pace * 0.1 + p.defending * 0.15;
    }
    if (p.position == "Defender") {
        return p.defending * 0.35 + p.physical * 0.2 + p.pace * 0.15 + p.passing * 0.2 + p.shooting * 0.1;
    }
    if (p.position == "Midfielder") {
        return p.passing * 0.3 + p.shooting * 0.15 + p.physical * 0.15 + p.pace * 0.15 + p.defending * 0.25;
    }
    return p.shooting * 0.35 + p.pace * 0.25 + p.physical * 0.15 + p.passing * 0.15 + p.defending * 0.1;
}

static double team_base_strength(const TeamSummary& team) {
    if (team.selected_xi.empty()) return 1.0;
    double total = 0.0;
    for (const auto& p : team.selected_xi) total += player_overall(p);
    return total / static_cast<double>(team.selected_xi.size());
}

static const PlayerSummary* pick_weighted(const std::vector<PlayerSummary>& players, std::mt19937& rng, const std::string& mode) {
    if (players.empty()) return nullptr;
    std::vector<int> weights;
    weights.reserve(players.size());
    for (const auto& p : players) {
        int w = 1;
        if (mode == "scorer") {
            if (p.position == "Forward") w = p.shooting * 3 + p.pace;
            else if (p.position == "Midfielder") w = p.shooting * 2 + p.passing;
            else if (p.position == "Defender") w = p.shooting + p.physical;
            else w = std::max(1, p.shooting / 2);
        } else if (mode == "assist") {
            w = p.passing * 2 + p.pace;
        } else if (mode == "discipline") {
            w = p.physical + p.defending;
        } else {
            w = std::max(1, static_cast<int>(std::round(player_overall(p))));
        }
        weights.push_back(std::max(1, w));
    }
    std::discrete_distribution<int> dist(weights.begin(), weights.end());
    return &players[dist(rng)];
}

static const PlayerSummary* pick_assister(const std::vector<PlayerSummary>& players, const std::string& scorer_id, std::mt19937& rng) {
    std::vector<PlayerSummary> candidates;
    for (const auto& p : players) {
        if (p.id != scorer_id && p.position != "Goalkeeper") candidates.push_back(p);
    }
    if (candidates.empty()) return nullptr;
    return pick_weighted(candidates, rng, "assist");
}

static double injury_chance(const TeamSummary& team, const PlayerSummary& player) {
    double intensity = static_cast<double>(team.training_intensity);
    double medical = static_cast<double>(team.medical_level);
    double pitch = static_cast<double>(team.pitch_quality);
    double age_mod = player.age >= 30 ? 0.01 : 0.0;
    double v = 0.015 + intensity * 0.01 - medical * 0.004 - pitch * 0.002 + age_mod;
    return std::max(0.005, v);
}

static void apply_table_result(WeekTableRow& home, WeekTableRow& away, const MatchOutput& result) {
    home.goals_for += result.home_goals;
    home.goals_against += result.away_goals;
    away.goals_for += result.away_goals;
    away.goals_against += result.home_goals;
    home.goal_difference = home.goals_for - home.goals_against;
    away.goal_difference = away.goals_for - away.goals_against;
    if (result.home_goals > result.away_goals) {
        home.wins += 1;
        home.points += 3;
        away.losses += 1;
    } else if (result.home_goals < result.away_goals) {
        away.wins += 1;
        away.points += 3;
        home.losses += 1;
    } else {
        home.draws += 1;
        away.draws += 1;
        home.points += 1;
        away.points += 1;
    }
}

} // namespace

MatchOutput simulate_match(const MatchInput& input) {
    MatchOutput output;
    output.home_team = input.home.name;
    output.away_team = input.away.name;

    std::mt19937 rng;
    if (input.has_seed) rng.seed(input.seed);
    else rng.seed(std::random_device{}());

    double home_base = team_base_strength(input.home) * (1.08 + input.home.pitch_quality * 0.005);
    double away_base = team_base_strength(input.away);

    add_event(output, 0, "Kick Off", "", "", "", "Kick off. " + input.home.name + " versus " + input.away.name + ".");

    std::uniform_real_distribution<double> prob(0.0, 1.0);
    std::uniform_int_distribution<int> stoppage_dist(2, 5);
    std::uniform_int_distribution<int> injury_weeks_dist(1, 5);

    int home_red_count = 0;
    int away_red_count = 0;

    for (int minute = 1; minute <= 90; ++minute) {
        if (minute == 46) {
            std::ostringstream ss;
            ss << "Half time. " << output.home_team << ' ' << output.home_goals << ", " << output.away_team << ' ' << output.away_goals << '.';
            add_event(output, 45, "Half Time", "", "", "", ss.str());
        }

        if (prob(rng) > 0.30) continue;

        double home_chance = home_base / std::max(1.0, away_base);
        double away_chance = away_base / std::max(1.0, home_base);
        double total = home_chance + away_chance;
        bool is_home = prob(rng) < (home_chance / total);

        const TeamSummary& atk_team = is_home ? input.home : input.away;
        const TeamSummary& def_team = is_home ? input.away : input.home;
        const auto& atk_xi = is_home ? input.home.selected_xi : input.away.selected_xi;
        const auto& def_xi = is_home ? input.away.selected_xi : input.home.selected_xi;

        double roll = prob(rng);
        if (roll < 0.15) {
            if (is_home) output.home_shots++; else output.away_shots++;
            double goal_chance = 0.33 + (is_home ? home_base : away_base) * 0.01;
            if (prob(rng) < goal_chance) {
                bool is_penalty = prob(rng) < 0.08;
                const PlayerSummary* scorer = pick_weighted(atk_xi, rng, "scorer");
                if (is_penalty) {
                    if (prob(rng) < 0.78) {
                        if (is_home) { output.home_goals++; output.home_on_target++; }
                        else { output.away_goals++; output.away_on_target++; }
                        std::ostringstream ss;
                        ss << minute << "' Goal for " << atk_team.name << ". Scorer: " << (scorer ? scorer->name : "Unknown")
                           << ". Score: " << output.home_team << ' ' << output.home_goals << ", " << output.away_team << ' ' << output.away_goals << '.';
                        add_event(output, minute, "Penalty Scored", atk_team.name, scorer ? scorer->name : "", "", ss.str());
                    } else {
                        std::ostringstream ss;
                        ss << minute << "' Penalty missed by " << (scorer ? scorer->name : "Unknown") << " of " << atk_team.name << '.';
                        add_event(output, minute, "Penalty Missed", atk_team.name, scorer ? scorer->name : "", "", ss.str());
                    }
                } else {
                    const PlayerSummary* assister = scorer ? pick_assister(atk_xi, scorer->id, rng) : nullptr;
                    if (is_home) { output.home_goals++; output.home_on_target++; }
                    else { output.away_goals++; output.away_on_target++; }
                    std::ostringstream ss;
                    ss << minute << "' Goal for " << atk_team.name << ". Scorer: " << (scorer ? scorer->name : "Unknown") << '.';
                    if (assister) ss << " Assist by " << assister->name << '.';
                    ss << " Score: " << output.home_team << ' ' << output.home_goals << ", " << output.away_team << ' ' << output.away_goals << '.';
                    add_event(output, minute, "Goal", atk_team.name, scorer ? scorer->name : "", assister ? assister->name : "", ss.str());
                }
            } else {
                const PlayerSummary* shooter = pick_weighted(atk_xi, rng, "scorer");
                if (prob(rng) < 0.5) {
                    if (is_home) output.home_on_target++; else output.away_on_target++;
                    std::ostringstream ss;
                    ss << minute << "' " << (shooter ? shooter->name : "A player") << " has a shot saved for " << atk_team.name << '.';
                    add_event(output, minute, "Shot Saved", atk_team.name, shooter ? shooter->name : "", "", ss.str());
                } else {
                    std::ostringstream ss;
                    ss << minute << "' " << (shooter ? shooter->name : "A player") << " shoots wide for " << atk_team.name << '.';
                    add_event(output, minute, "Shot Wide", atk_team.name, shooter ? shooter->name : "", "", ss.str());
                }
            }
        } else if (roll < 0.30) {
            if (is_home) output.home_corners++; else output.away_corners++;
            add_event(output, minute, "Corner", atk_team.name, "", "", std::to_string(minute) + "' Corner to " + atk_team.name + ".");
        } else if (roll < 0.55) {
            const PlayerSummary* fouler = pick_weighted(def_xi, rng, "discipline");
            if (is_home) output.away_fouls++; else output.home_fouls++;
            add_event(output, minute, "Foul", def_team.name, fouler ? fouler->name : "", "", std::to_string(minute) + "' Foul by " + (fouler ? fouler->name : std::string("a player")) + " of " + def_team.name + ".");
            double card_roll = prob(rng);
            if (card_roll < 0.15 && fouler) {
                if (is_home) output.away_yellows++; else output.home_yellows++;
                add_event(output, minute, "Yellow Card", def_team.name, fouler->name, "", std::to_string(minute) + "' Yellow card for " + fouler->name + " of " + def_team.name + ".");
            } else if (card_roll < 0.03 && fouler) {
                if (is_home) { output.away_reds++; away_red_count++; }
                else { output.home_reds++; home_red_count++; }
                int count = is_home ? 11 - away_red_count : 11 - home_red_count;
                add_event(output, minute, "Red Card", def_team.name, fouler->name, "", std::to_string(minute) + "' Red card for " + fouler->name + " of " + def_team.name + ". " + def_team.name + " are down to " + std::to_string(count) + " men.");
            }
        } else if (roll < 0.62) {
            const PlayerSummary* player = pick_weighted(atk_xi, rng, "default");
            if (player && prob(rng) < injury_chance(atk_team, *player)) {
                int weeks = injury_weeks_dist(rng);
                add_event(output, minute, "Injury", atk_team.name, player->name, "", std::to_string(minute) + "' Injury concern for " + atk_team.name + ". " + player->name + " may miss " + std::to_string(weeks) + " weeks.");
            }
        }
    }

    int stoppage = stoppage_dist(rng);
    for (int minute = 91; minute < 91 + stoppage; ++minute) {
        if (prob(rng) < 0.12) {
            bool is_home = prob(rng) < 0.5;
            const TeamSummary& atk_team = is_home ? input.home : input.away;
            const auto& atk_xi = is_home ? input.home.selected_xi : input.away.selected_xi;
            if (is_home) output.home_shots++; else output.away_shots++;
            if (prob(rng) < 0.25) {
                const PlayerSummary* scorer = pick_weighted(atk_xi, rng, "scorer");
                if (is_home) { output.home_goals++; output.home_on_target++; }
                else { output.away_goals++; output.away_on_target++; }
                std::ostringstream ss;
                ss << "90+" << (minute - 90) << "' Goal for " << atk_team.name << ". Scorer: " << (scorer ? scorer->name : "Unknown")
                   << ". Score: " << output.home_team << ' ' << output.home_goals << ", " << output.away_team << ' ' << output.away_goals << '.';
                add_event(output, minute, "Goal", atk_team.name, scorer ? scorer->name : "", "", ss.str());
            }
        }
    }

    add_event(output, 90, "Full Time", "", "", "", "Full time. " + output.home_team + " " + std::to_string(output.home_goals) + ", " + output.away_team + " " + std::to_string(output.away_goals) + ".");

    std::uniform_real_distribution<double> fill_pct(0.4, 0.9);
    output.attendance = static_cast<int>(input.home.stadium_capacity * fill_pct(rng));
    return output;
}

WeekSimulationOutput simulate_week(const WeekSimulationInput& input) {
    WeekSimulationOutput out;
    std::map<std::string, WeekTableRow> table_map;

    for (const auto& fixture : input.fixtures) {
        MatchInput match_input;
        match_input.home = fixture.home;
        match_input.away = fixture.away;
        match_input.seed = fixture.seed;
        match_input.has_seed = fixture.has_seed;

        MatchOutput match_result = simulate_match(match_input);
        WeekFixtureResult fixture_result;
        fixture_result.fixture_id = fixture.fixture_id;
        fixture_result.competition_id = fixture.competition_id;
        fixture_result.stage = fixture.stage;
        fixture_result.week = fixture.week;
        fixture_result.match = match_result;
        out.results.push_back(fixture_result);

        auto& home_row = table_map[fixture.home.id];
        if (home_row.club_id.empty()) {
            home_row.club_id = fixture.home.id;
            home_row.club_name = fixture.home.name;
        }
        auto& away_row = table_map[fixture.away.id];
        if (away_row.club_id.empty()) {
            away_row.club_id = fixture.away.id;
            away_row.club_name = fixture.away.name;
        }
        apply_table_result(home_row, away_row, match_result);
    }

    for (auto& kv : table_map) {
        out.table.push_back(kv.second);
    }
    std::sort(out.table.begin(), out.table.end(), [](const WeekTableRow& a, const WeekTableRow& b) {
        if (a.points != b.points) return a.points > b.points;
        if (a.goal_difference != b.goal_difference) return a.goal_difference > b.goal_difference;
        if (a.goals_for != b.goals_for) return a.goals_for > b.goals_for;
        return a.club_name < b.club_name;
    });
    return out;
}

} // namespace fm

File: match_engine.py
"""Match simulation engine for Football Manager 26."""

import random
from models import (
    Club, MatchEvent, MatchResult, EventType, Position,
    Mentality, FORMATION_SLOTS,
)

GOAL_TEMPLATES = [
    "{player} scores for {team}! A well-taken goal.",
    "GOAL! {player} finds the back of the net for {team}!",
    "{player} with a brilliant finish! The ball is in the net!",
    "It's a goal! {player} puts {team} ahead!",
    "{player} scores! What a moment for {team}!",
    "Clinical finish from {player}! {team} celebrate!",
    "The ball hits the net! {player} has scored for {team}!",
]
ASSIST_TEMPLATES = [
    " Assisted by {assist}.",
    " Great pass from {assist} to set up the goal.",
    " {assist} with the assist.",
    " Brilliant play from {assist} to create the chance.",
]
SHOT_SAVED_TEMPLATES = [
    "{player} shoots but it's saved by the goalkeeper.",
    "Good save! {player}'s shot is kept out.",
    "{player} forces a save from the keeper.",
    "The goalkeeper gets down well to stop {player}'s effort.",
]
SHOT_WIDE_TEMPLATES = [
    "{player} shoots but it goes wide of the post.",
    "{player}'s effort sails over the bar.",
    "Off target! {player} can't find the goal.",
    "{player} blazes it over from a good position.",
]
FOUL_TEMPLATES = [
    "Foul by {player} of {team}.",
    "{player} commits a foul. Free kick awarded.",
    "The referee blows for a foul on {player}.",
]
YELLOW_TEMPLATES = [
    "Yellow card! {player} of {team} is booked.",
    "{player} receives a yellow card for that challenge.",
    "The referee shows yellow to {player}.",
]
RED_TEMPLATES = [
    "RED CARD! {player} of {team} is sent off!",
    "{player} receives a straight red card! {team} down to {count} men!",
    "The referee shows red! {player} must leave the field!",
]
CORNER_TEMPLATES = [
    "Corner kick to {team}.",
    "{team} win a corner.",
]
INJURY_TEMPLATES = [
    "{player} goes down injured. The physio is called on.",
    "Concern for {team} as {player} appears to be hurt.",
]
PENALTY_SCORED_TEMPLATES = [
    "PENALTY SCORED! {player} sends the keeper the wrong way!",
    "{player} converts from the spot! Cool as you like!",
]
PENALTY_MISSED_TEMPLATES = [
    "PENALTY MISSED! {player} puts it wide!",
    "The goalkeeper saves the penalty from {player}!",
    "{player} hits the post from the spot!",
]


def get_starting_eleven(club):
    formation = club.tactic.formation
    slots = FORMATION_SLOTS[formation]
    available = [p for p in club.players if p.is_available]
    starting = []
    for pos, count in slots.items():
        candidates = sorted([p for p in available if p.position == pos and p not in starting], key=lambda p: p.overall, reverse=True)
        for i in range(count):
            if i < len(candidates):
                starting.append(candidates[i])
            else:
                remaining = [p for p in available if p not in starting]
                if remaining:
                    remaining.sort(key=lambda p: p.overall, reverse=True)
                    starting.append(remaining[0])
    return starting[:11]


def calculate_team_strength(club, starting_eleven=None):
    if starting_eleven is None:
        starting_eleven = get_starting_eleven(club)
    if not starting_eleven:
        return 1.0, 1.0, 1.0

    base = sum(p.overall for p in starting_eleven) / len(starting_eleven)
    attack_mod = 1.0
    defence_mod = 1.0
    mentality = club.tactic.mentality
    if mentality == Mentality.DEFENSIVE:
        attack_mod = 0.85
        defence_mod = 1.15
    elif mentality == Mentality.CAUTIOUS:
        attack_mod = 0.92
        defence_mod = 1.08
    elif mentality == Mentality.ATTACKING:
        attack_mod = 1.12
        defence_mod = 0.88
    elif mentality == Mentality.ALL_OUT:
        attack_mod = 1.25
        defence_mod = 0.75
    return base, attack_mod, defence_mod


def _pick_scorer(players):
    if not players:
        return None
    weights = []
    for p in players:
        if p.position == Position.FWD:
            w = p.shooting * 3 + p.pace
        elif p.position == Position.MID:
            w = p.shooting * 2 + p.passing
        elif p.position == Position.DEF:
            w = p.shooting + p.physical
        else:
            w = 1
        weights.append(max(1, w))
    return random.choices(players, weights=weights, k=1)[0]


def _pick_assister(players, scorer):
    candidates = [p for p in players if p != scorer and p.position != Position.GK]
    if not candidates:
        return None
    weights = [max(1, p.passing * 2 + p.pace) for p in candidates]
    return random.choices(candidates, weights=weights, k=1)[0]


def _pick_any(players):
    if not players:
        return None
    outfield = [p for p in players if p.position != Position.GK]
    pool = outfield if outfield else players
    return random.choice(pool) if pool else None


def _injury_chance(club, player):
    infra = club.infrastructure
    intensity = infra.training.intensity
    medical = infra.training.medical_level
    pitch = infra.stadium.pitch_quality
    age_mod = 0.01 if player.age >= 30 else 0
    return max(0.005, 0.015 + intensity * 0.01 - medical * 0.004 - pitch * 0.002 + age_mod)


def simulate_match(home_club, away_club):
    home_xi = get_starting_eleven(home_club)
    away_xi = get_starting_eleven(away_club)
    home_base, home_atk, home_def = calculate_team_strength(home_club, home_xi)
    away_base, away_atk, away_def = calculate_team_strength(away_club, away_xi)

    home_base *= 1.08 + (home_club.infrastructure.stadium.pitch_quality * 0.005)

    events = []
    home_goals = 0
    away_goals = 0
    stats = {
        "home_shots": 0, "away_shots": 0,
        "home_on_target": 0, "away_on_target": 0,
        "home_corners": 0, "away_corners": 0,
        "home_fouls": 0, "away_fouls": 0,
        "home_yellows": 0, "away_yellows": 0,
        "home_reds": 0, "away_reds": 0,
    }
    home_red_count = 0
    away_red_count = 0

    events.append(MatchEvent(0, EventType.KICK_OFF, commentary=f"Kick off. {home_club.name} versus {away_club.name}."))

    for minute in range(1, 91):
        if minute == 46:
            events.append(MatchEvent(45, EventType.HALF_TIME, commentary=f"Half time. {home_club.name} {home_goals}, {away_club.name} {away_goals}."))

        if random.random() > 0.30:
            continue

        home_chance = home_base * home_atk / max(1, away_base * away_def)
        away_chance = away_base * away_atk / max(1, home_base * home_def)
        total = home_chance + away_chance
        is_home = random.random() < (home_chance / total)

        atk_team = home_club if is_home else away_club
        def_team = away_club if is_home else home_club
        atk_xi = home_xi if is_home else away_xi
        def_xi = away_xi if is_home else home_xi
        prefix = "home" if is_home else "away"

        if is_home and home_red_count > 0:
            home_base *= 0.95
        elif not is_home and away_red_count > 0:
            away_base *= 0.95

        roll = random.random()

        if roll < 0.15:
            stats[f"{prefix}_shots"] += 1
            goal_chance = 0.33 + (home_base if is_home else away_base) * 0.01
            if random.random() < goal_chance:
                is_penalty = random.random() < 0.08
                scorer = _pick_scorer(atk_xi)
                if scorer:
                    scorer.appearances = max(scorer.appearances, 1)
                if is_penalty:
                    if random.random() < 0.78:
                        if is_home:
                            home_goals += 1
                        else:
                            away_goals += 1
                        stats[f"{prefix}_on_target"] += 1
                        if scorer:
                            scorer.goals += 1
                        commentary = f"{minute}' Goal for {atk_team.name}. Scorer: {(scorer.full_name if scorer else 'Unknown')}. Score: {home_club.name} {home_goals}, {away_club.name} {away_goals}."
                        events.append(MatchEvent(minute, EventType.PENALTY_SCORED, atk_team.name, scorer.full_name if scorer else "", commentary=commentary))
                    else:
                        commentary = f"{minute}' Penalty missed by {(scorer.full_name if scorer else 'Unknown')} of {atk_team.name}."
                        events.append(MatchEvent(minute, EventType.PENALTY_MISSED, atk_team.name, scorer.full_name if scorer else "", commentary=commentary))
                else:
                    assister = _pick_assister(atk_xi, scorer) if scorer else None
                    if is_home:
                        home_goals += 1
                    else:
                        away_goals += 1
                    stats[f"{prefix}_on_target"] += 1
                    if scorer:
                        scorer.goals += 1
                    assist_text = f" Assist by {assister.full_name}." if assister else ""
                    if assister:
                        assister.assists += 1
                        assister.appearances = max(assister.appearances, 1)
                    commentary = f"{minute}' Goal for {atk_team.name}. Scorer: {(scorer.full_name if scorer else 'Unknown')}.{assist_text} Score: {home_club.name} {home_goals}, {away_club.name} {away_goals}."
                    events.append(MatchEvent(minute, EventType.GOAL, atk_team.name, scorer.full_name if scorer else "", assister.full_name if assister else "", commentary))
            else:
                shooter = _pick_scorer(atk_xi)
                if random.random() < 0.5:
                    stats[f"{prefix}_on_target"] += 1
                    commentary = f"{minute}' {(shooter.full_name if shooter else 'A player')} has a shot saved for {atk_team.name}."
                    events.append(MatchEvent(minute, EventType.SHOT_SAVED, atk_team.name, shooter.full_name if shooter else "", commentary=commentary))
                else:
                    commentary = f"{minute}' {(shooter.full_name if shooter else 'A player')} shoots wide for {atk_team.name}."
                    events.append(MatchEvent(minute, EventType.SHOT_WIDE, atk_team.name, shooter.full_name if shooter else "", commentary=commentary))

        elif roll < 0.30:
            stats[f"{prefix}_corners"] += 1
            commentary = f"{minute}' Corner to {atk_team.name}."
            events.append(MatchEvent(minute, EventType.CORNER, atk_team.name, commentary=commentary))

        elif roll < 0.55:
            fouler = _pick_any(def_xi)
            stats[f"{'away' if is_home else 'home'}_fouls"] += 1
            commentary = f"{minute}' Foul by {(fouler.full_name if fouler else 'a player')} of {def_team.name}."
            events.append(MatchEvent(minute, EventType.FOUL, def_team.name, fouler.full_name if fouler else "", commentary=commentary))
            card_roll = random.random()
            if card_roll < 0.15 and fouler:
                opp = "away" if is_home else "home"
                stats[f"{opp}_yellows"] += 1
                fouler.yellow_cards += 1
                commentary = f"{minute}' Yellow card for {fouler.full_name} of {def_team.name}."
                events.append(MatchEvent(minute, EventType.YELLOW_CARD, def_team.name, fouler.full_name, commentary=commentary))
                if fouler.yellow_cards >= 5:
                    fouler.suspended_matches = 1
            elif card_roll < 0.03 and fouler:
                opp = "away" if is_home else "home"
                stats[f"{opp}_reds"] += 1
                fouler.red_cards += 1
                if is_home:
                    away_red_count += 1
                    count = 11 - away_red_count
                else:
                    home_red_count += 1
                    count = 11 - home_red_count
                fouler.suspended_matches = 3
                commentary = f"{minute}' Red card for {fouler.full_name} of {def_team.name}. {def_team.name} are down to {count} men."
                events.append(MatchEvent(minute, EventType.RED_CARD, def_team.name, fouler.full_name, commentary=commentary))

        elif roll < 0.62:
            player = _pick_any(atk_xi)
            if player and random.random() < _injury_chance(atk_team, player):
                player.injured_weeks = random.randint(1, 5)
                commentary = f"{minute}' Injury concern for {atk_team.name}. {player.full_name} is hurt and may miss {player.injured_weeks} weeks."
                events.append(MatchEvent(minute, EventType.INJURY, atk_team.name, player.full_name, commentary=commentary))

    stoppage = random.randint(2, 5)
    for minute in range(91, 91 + stoppage):
        if random.random() < 0.12:
            is_home = random.random() < 0.5
            atk_team = home_club if is_home else away_club
            atk_xi = home_xi if is_home else away_xi
            prefix = "home" if is_home else "away"
            stats[f"{prefix}_shots"] += 1
            if random.random() < 0.25:
                scorer = _pick_scorer(atk_xi)
                if is_home:
                    home_goals += 1
                else:
                    away_goals += 1
                stats[f"{prefix}_on_target"] += 1
                if scorer:
                    scorer.goals += 1
                commentary = f"90+{minute-90}' Goal for {atk_team.name}. Scorer: {(scorer.full_name if scorer else 'Unknown')}. Score: {home_club.name} {home_goals}, {away_club.name} {away_goals}."
                events.append(MatchEvent(minute, EventType.GOAL, atk_team.name, scorer.full_name if scorer else "", commentary=commentary))

    events.append(MatchEvent(90, EventType.FULL_TIME, commentary=f"Full time. {home_club.name} {home_goals}, {away_club.name} {away_goals}."))

    for p in home_xi + away_xi:
        p.appearances += 1

    base_cap = home_club.stadium_capacity
    fill_pct = random.uniform(0.4, 0.9)
    attendance = int(base_cap * fill_pct)

    return MatchResult(
        home_team=home_club.name,
        away_team=away_club.name,
        home_goals=home_goals,
        away_goals=away_goals,
        events=events,
        attendance=attendance,
        **stats,
    )
